Add "SimplyE" and "Open eBooks" as subprojects

twaddington opened the pull request:
This change adds the two subprojects `simplified-app-simplye` and `simplified-app-openebooks`, which were migrated using the following script:

https://gist.github.com/twaddington/45be011e6d65f230aa8b26883c91c703

There are several advantages to this change.

- Redundant build configuration has been consolidated into the root `build.gradle` file.
- Dependency versions can be pinned across all projects within `libraries.gradle`.
- App versions will track Core and will always be up to date.
- Refactoring across projects becomes simpler and requires only a single pull-request.
- Less app versions for QA to track.

If this change is adopted, the GitHub repositories for [SimplyE](https://github.com/NYPL-Simplified/Simplified-Android-SimplyE) and [Open eBooks](https://github.com/NYPL-Simplified/Simplified-Android-Open-eBooks) can be deleted.

Additionally, major improvements have been made to the build scripts. For example, app secrets will now be copied into builds _automatically_ so long as you define the source directory in your `~/.gradle/gradle.properties` config.

You'll need to make sure your config is updated to include something like the following:

```properties
>>> ~/.gradle/gradle.properties

# Required for publishing to Maven Central
#
mavenCentralUsername=[redacted]
mavenCentralPassword=[redacted]

# Required for building with Adobe DRM or publishing to the NYPL Nexus
#
org.librarysimplified.nexus.depend=true
org.librarysimplified.nexus.username=[redacted]
org.librarysimplified.nexus.password=[redacted]

# Required for signing release builds
#
org.librarysimplified.keyAlias=nypl
org.librarysimplified.keyPassword=[redacted]
org.librarysimplified.storePassword=[redacted]

# Required for copying app secrets into the build at compile time
#
org.librarysimplified.app.assets.openebooks=\
    /path/to/Certificates/OpenEBooks/Android/
org.librarysimplified.app.assets.simplye=\
    /path/to/Certificates/SimplyE/Android/

# Skip the tasks of any projects that require drm, unless enabled
#
org.librarysimplified.drm.enabled=true

# Miscellaneous
#
#org.librarysimplified.no_signing=true
#org.librarysimplified.publish.directory=\
#    /path/to/foo/
```

twaddington commented:
@io7m this is pretty solid now so please take a look when you have some time.

io7m commented on buildSrc/src/main/java/org/librarysimplified/gradle/RequiredAssetsTask.kt:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  private val strictMode by lazy {
+    project.findProperty(STRICT_MODE)?.toString()?.toBoolean() ?: false
```
I'm not sure I like this being off by default, and this was one of the major reasons that SimplyE was in a separate repos: Someone building SimplyE in that other repos is _required_ to have set everything up properly and the build will fail early if they haven't. This new setup would seem to allow them to build a broken application with missing files and then find out about it at runtime.

twaddington replied:
Followed up on this below!

io7m commented on simplified-app-openebooks/google-services.json:
```diff
@@ -0,0 +1,40 @@
+      ],
+      "api_key": [
+        {
+          "current_key": "AIzaSyBzGt9lNXa3vO0zJccs_qlPQrwDhEYCR6c"
```
Is this really not a secret?

twaddington replied:
Yep! You can grab any published APK file and extract this key with no special tools. Any web or mobile app using the Firebase API will have an API key like this.

Google is pretty explicit here (emphasis mine):

> The content of the Firebase config file or object **_is considered public_**, including the app's platform-specific ID (iOS bundle ID or Android package name) and the Firebase project-specific values, like the API Key, project ID, Realtime Database URL, and Storage bucket name.

Source https://firebase.google.com/docs/projects/learn-more#config-files-objects

See also [here](https://groups.google.com/g/firebase-talk/c/bamCgTDajkw/m/uVEJXjtiBwAJ?pli=1) and [here](https://stackoverflow.com/questions/37482366/is-it-safe-to-expose-firebase-api).

io7m commented on simplified-app-openebooks/src/main/res/values/features.xml:
```diff
@@ -0,0 +1,16 @@
+    <color name="feature_secondary_color">#7baf38</color>
+    <!-- The default Adobe vendor ID -->
+    <string name="feature_adobe_vendor_id">OEI</string>
+</resources>
```
I'm not completely sure if any of these `feature_` settings are still used. These were pre-LFA, even...

twaddington replied:
Hah, yeah, I had no idea what these were for. Happy to delete them if they're unused.

io7m commented on simplified-app-simplye/google-services.json:
```diff
@@ -0,0 +1,40 @@
+      ],
+      "api_key": [
+        {
+          "current_key": "AIzaSyBkM8WKKf-LCL0dMldtvKx5dj6eVAmpCew"
```
Ditto!

twaddington replied:
See above! 😅

io7m commented:
I'm fine with this in general, but I think the behaviour should change slightly:

If you don't explicitly ask to build with DRM, you shouldn't get APKs for SimplyE or Open eBooks at all because they'll obviously be non-working. If you _do_ ask to build with DRM, then you get what we call strict mode right now, and APKs of all three apps should be built. This should give random-github-passer-by-friendly behaviour: If some random person checks out the repository and runs `gradlew`, they'll get a vanilla APK without having to do any kind of configuration. Developers that are actually working on SimplyE set a single property somewhere and get working APKs of all three apps.

twaddington commented:
> If you don't explicitly ask to build with DRM, you shouldn't get APKs for SimplyE or Open eBooks at all because they'll obviously be non-working.

I can make the project work that way, for sure.

Typically in a multi-module gradle project like this you'd specify explicitly which submodule to build:

```
# Just vanilla and dependencies
#
$ ./gradlew :simplified-app-vanilla:assemble

# SimplyE and Open Ebooks
#
$ ./gradlew :simplified-app-simplye:assemble :simplified-app-openebooks:assemble
```

That's obviously not as straightforward as just running `./gradlew assemble` on the whole shebang.

> If you _do_ ask to build with DRM, then you should get what we call strict mode right now, and APKs of all three apps should be built.

I'll rename "strictmode" to something like `org.librarysimplified.drm.enabled=true` and update the build to omit those two apps unless that property is set. That should be fairly straightforward and I think will work exactly as you suggest.

> This should give random-github-passer-by-friendly behaviour: If some random person checks out the repository and runs `gradlew`, they'll get a vanilla APK without having to do any kind of configuration. Developers that are actually working on SimplyE set a single property somewhere (well, a single property and a pile of other properties for credentials 🙂) and get working APKs of all three apps.

Yeah, this was what I had in mind as the original goal. First, optimize for the vanilla user who checks out the repo for the first time and has none of the private credentials. However, I also wanted to make sure NYPL developer couldn't accidentally generate a broken version of SimplyE without all the necessary secrets. Lots of different ways to slice this!

io7m commented:
Sounds good!

twaddington commented:
@io7m I believe 5eaa815fc5b2ea1a0a7f565cc2a2586b466349d5 should resolve your major concerns.

io7m commented:
This seems fine to me. :+1:

Do you think this is ready to merge now, or is there anything else you want to do with it? I'd like to get this merged sooner rather than later as it affects how our builds are published.

io7m commented:
Actually, as a random github passer-by without credentials, I get this:

```
$ ./gradlew clean assemble
...
...
...
* What went wrong:
Execution failed for task ':simplified-tests-sandbox:preReleaseCheck'.
> Keystore file '/home/rm-work/git/com.github/NYPL-Simplified/Simplified-Android-Core/release.jks' must exist to sign release builds
```

Slightly irritatingly, it happens right at the end of the build.

twaddington commented:
> Actually, as a random github passer-by without credentials, I get this:
> 
> ```
> $ ./gradlew clean assemble
> ...
> ...
> ...
> * What went wrong:
> Execution failed for task ':simplified-tests-sandbox:preReleaseCheck'.
> > Keystore file '/home/rm-work/git/com.github/NYPL-Simplified/Simplified-Android-Core/release.jks' must exist to sign release builds
> ```
> 
> Slightly irritatingly, it happens right at the end of the build.

Yeah, that's expected. You will need a keystore to sign release builds. That's mostly unavoidable but you can run `./gradlew assembleDebug` to avoid generating that build type.

twaddington commented:
Talked with @io7m about this and I think we're good to go. I'm gonna do one final look over and then will have a couple small follow-on pulls that fix a couple other things I noticed.

io7m commented:
:rocket:
